File: test/org/jetbrains/plugins/scala/lang/completion/CompletionTestBase.java
package org.jetbrains.plugins.scala.lang.completion;

import com.intellij.codeInsight.CodeInsightActionHandler;
import com.intellij.codeInsight.completion.CompletionData;
import com.intellij.codeInsight.completion.CompletionUtil;
import com.intellij.codeInsight.completion.CompletionVariant;
import com.intellij.codeInsight.completion.actions.CodeCompletionAction;
import com.intellij.codeInsight.completion.impl.CamelHumpMatcher;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.fileEditor.OpenFileDescriptor;
import com.intellij.openapi.util.InvalidDataException;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiReference;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.plugins.scala.lang.actions.ActionTestBase;
import org.jetbrains.plugins.scala.util.TestUtils;

import java.io.IOException;
import java.util.*;

/**
 * @author Alexander Podkhalyuzin
 */
public abstract class CompletionTestBase extends ActionTestBase {
  protected Editor myEditor;
  protected FileEditorManager myFileEditorManager;
  protected PsiFile myFile;
  protected int myOffset;

  public CompletionTestBase(String path) {
    super(path);
  }

  protected CodeInsightActionHandler getCompetionHandler() {
    CodeCompletionAction action = new CodeCompletionAction();
    return action.getHandler();
  }

  protected String processFile(final PsiFile file) throws IncorrectOperationException, InvalidDataException, IOException {
    String result = "";
    String fileText = file.getText();
    int offset = fileText.indexOf(CARET_MARKER);
    fileText = removeMarker(fileText);
    myFile = createFile(fileText);
    myFileEditorManager = FileEditorManager.getInstance(myProject);
    myEditor = myFileEditorManager.openTextEditor(new OpenFileDescriptor(myProject, myFile.getVirtualFile(), 0), false);
    myEditor.getCaretModel().moveToOffset(offset);
    myOffset = myEditor.getCaretModel().getOffset();

    final CodeInsightActionHandler handler = getCompetionHandler();
    CompletionData data = CompletionUtil.getCompletionDataByElement(myFile);
    LookupElement[] items = getAcceptableItems(data);

    try {
      /*performAction(project, new Runnable() {
        public void run() {
          handler.invoke(project, myEditor, myFile);
        }
      });*/

      offset = myEditor.getCaretModel().getOffset();

      if (items.length > 0) {
        Arrays.sort(items);
        result = "";
        for (LookupElement item : items) {
          result = result + "\n" + item.getLookupString();
        }
        result = result.trim();
      }

    } finally {
      myFileEditorManager.closeFile(myFile.getVirtualFile());
      myEditor = null;
    }
    return result;
  }

  protected PsiFile createFile(String fileText) throws IncorrectOperationException {
    return TestUtils.createPseudoPhysicalScalaFile(myProject, fileText);
  }

  protected abstract LookupElement[] getAcceptableItems(CompletionData data) throws IncorrectOperationException;


  /**
   * retrurns acceptable variant for this completion
   *
   * @param completionData
   * @return
   */
  protected LookupElement[] getAcceptableItemsImpl(CompletionData completionData) throws IncorrectOperationException {

    final Set<LookupElement> lookupSet = new LinkedHashSet<LookupElement>();
    final PsiElement elem = myFile.findElementAt(myOffset);

    /**
     * Create fake file with dummy element
     */
    String newFileText = myFile.getText().substring(0, myOffset) + "IntellijIdeaRulezzz" +
            myFile.getText().substring(myOffset);
    /**
     * Hack for IDEA completion
     */
    PsiFile newFile = createFile(newFileText);
    PsiElement insertedElement = newFile.findElementAt(myOffset + 1);
    String prefix = newFile.getText().substring(insertedElement.getTextRange().getStartOffset(), myOffset);
    if (lookupSet.size() == 0) {
      final PsiReference ref = newFile.findReferenceAt(myOffset + 1);
      if (addKeywords(ref)) {
        final Set<CompletionVariant> keywordVariants = new HashSet<CompletionVariant>();
        completionData.addKeywordVariants(keywordVariants, insertedElement, newFile);
        completionData.completeKeywordsBySet(lookupSet, keywordVariants, insertedElement, new CamelHumpMatcher(prefix), newFile);
      }
      //todo: add variant for reference completion
    }

    ArrayList<LookupElement> lookupItems = new ArrayList<LookupElement>();
    final LookupElement[] items = lookupSet.toArray(new LookupElement[lookupSet.size()]);
    for (LookupElement item : items) if (item.getLookupString().startsWith(prefix)){
      lookupItems.add(item);
    }

    return lookupItems.toArray(new LookupElement[lookupItems.size()]);

  }

  public String transform(String testName, String[] data) throws Exception {
    setSettings();
    String fileText = data[0];
    final PsiFile psiFile = createFile(fileText);
    String result = processFile(psiFile);
    System.out.println("------------------------ " + testName + " ------------------------");
    System.out.println(result);
    System.out.println("");
    return result;
  }

  protected abstract boolean addKeywords(PsiReference ref);

  protected abstract boolean addReferenceVariants(PsiReference ref);
}

File: test/org/jetbrains/plugins/scala/lang/completion/CompletionTestBase.scala
package org.jetbrains.plugins.scala
package lang.completion

import java.io.File

import com.intellij.codeInsight.completion.{CodeCompletionHandlerBase, CompletionType}
import com.intellij.codeInsight.lookup.impl.LookupImpl
import com.intellij.codeInsight.lookup.{LookupElement, LookupManager}
import com.intellij.openapi.fileEditor.{FileEditorManager, OpenFileDescriptor}
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.{CharsetToolkit, LocalFileSystem, VirtualFile}
import org.jetbrains.plugins.scala.base.ScalaLightPlatformCodeInsightTestCaseAdapter
import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes
import org.jetbrains.plugins.scala.lang.psi.api.ScalaFile

/**
 * User: Alexander Podkhalyuzin
 * Date: 23.09.2009
 */

abstract class CompletionTestBase extends ScalaLightPlatformCodeInsightTestCaseAdapter {
  protected val caretMarker = "/*caret*/"

  def folderPath: String = baseRootPath() + "completion/"
  def testFileExt: String = ".scala"

  protected def loadFile: (String, VirtualFile) = {
    val fileName = getTestName(false) + testFileExt
    val filePath = folderPath + fileName
    val file = LocalFileSystem.getInstance.findFileByPath(filePath.replace(File.separatorChar, '/'))
    assert(file != null, "file " + filePath + " not found")
    (fileName, file)
  }

  protected def loadAndSetFileText(filePath: String, file: VirtualFile): String = {
    val fileText = StringUtil.convertLineSeparators(FileUtil.loadFile(new File(file.getCanonicalPath), CharsetToolkit.UTF8))
    configureFromFileTextAdapter(filePath, fileText)
    fileText
  }

  protected def extractCaretOffset(fileText: String): Int = {
    val offset = fileText.indexOf(caretMarker)
    assert(offset != -1, "Not specified end marker in test case. Use /*caret*/ in scala file for this.")
    offset
  }

  /**
   * Open {@code file} in editor at {@code offset} position,
   * perform completion and return results
   * @param file VFS file to open
   * @param offset Caret position
   * @return Array of lookup strings
   */
  protected def getCompletionItems(file: VirtualFile, offset: Integer): Array[String] = {
    val fileEditorManager = FileEditorManager.getInstance(getProjectAdapter)
    val editor = fileEditorManager.openTextEditor(new OpenFileDescriptor(getProjectAdapter, file, offset), false)

    val completionType = if (getTestName(false).startsWith("Smart")) CompletionType.SMART else CompletionType.BASIC
    new CodeCompletionHandlerBase(completionType, false, false, true).invokeCompletion(getProjectAdapter, editor)
    val lookup: LookupImpl = LookupManager.getActiveLookup(editor).asInstanceOf[LookupImpl]

    if (lookup == null)
      Array.empty
    else
      lookup.getItems.toArray(LookupElement.EMPTY_ARRAY).map(_.getLookupString)
  }

  /**
   * Fetches last PSI element, checks if it is comment or not
   * If it is some kind of comment, treat it like an expected result string
   * If it's not, fail and return empty string
   * @return Expected result string
   */
  protected def getExpectedResult: String = {
    val scalaFile = getFileAdapter.asInstanceOf[ScalaFile]
    val lastPsi = scalaFile.findElementAt(scalaFile.getText.length - 1)
    val text = lastPsi.getText
    lastPsi.getNode.getElementType match {
      case ScalaTokenTypes.tLINE_COMMENT =>
        text.substring(2).trim
      case ScalaTokenTypes.tBLOCK_COMMENT | ScalaTokenTypes.tDOC_COMMENT =>
        text.substring(2, text.length - 2).trim
      case _ =>
        assert(assertion = false, "Test result must be in last comment statement.")
        ""
    }
  }

  protected def checkResult(got: Array[String], expected: String) {
    import junit.framework.Assert._
    val res = got.sortWith(_ < _).mkString("\n")
    assertEquals(expected, res.trim)
  }

  protected def doTest() {
    val (filePath, file) = loadFile
    val fileText = loadAndSetFileText(filePath, file)
    val offset = extractCaretOffset(fileText)
    val items = getCompletionItems(file, offset)
    val expected = getExpectedResult
    checkResult(items, expected)
  }
}